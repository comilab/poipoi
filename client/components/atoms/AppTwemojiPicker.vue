<template lang="pug">
TwemojiPicker.twemoji-picker(
  :emojiData="emojiData",
  :emojiGroups="EmojiGroups",
  :pickerWidth="pickerWidth",
  :searchEmojisFeat="true",
  :emojiPickerDisabled="disabled",
  @emojiUnicodeAdded="onSelect"
)
  template(#twemoji-picker-button)
    slot
</template>

<script lang="ts">
import { defineComponent, computed } from 'nuxt-composition-api'
import { PropType } from '@vue/composition-api'
// @ts-ignore
import { TwemojiPicker } from '@kevinfaguiar/vue-twemoji-picker'
import EmojiPack from '@kevinfaguiar/vue-twemoji-picker/src/interfaces/EmojiPack'
import EmojiAllData from '@kevinfaguiar/vue-twemoji-picker/emoji-data/ja/emoji-all-groups.json'
import EmojiGroups from '@kevinfaguiar/vue-twemoji-picker/emoji-data/emoji-groups.json'
import { useWindowSize } from '@vueuse/core'

export default defineComponent({
  components: {
    TwemojiPicker
  },
  props: {
    emojis: {
      type: Array as PropType<string[]>,
      required: false,
      default: null
    },
    disabled: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  setup (props, { emit }) {
    const { width } = useWindowSize()

    const emojiData = computed(() => {
      if (!props.emojis?.length) {
        return EmojiAllData
      }
      return (EmojiAllData as EmojiPack[])
        .map((pack) => {
          const emojiList = pack.emojiList.filter((emojiData) => {
            const emoji = emojiData.unicode
            return props.emojis.includes(emoji)
          })
          return {
            ...pack,
            emojiList
          }
        })
        .filter(pack => !!pack.emojiList.length)
    })

    const pickerWidth = computed(() => {
      if (width.value > 532) {
        return 500
      }
      return width.value - 32
    })

    const onSelect = (emoji: string) => {
      emit('select', emoji)
    }

    return {
      emojiData,
      pickerWidth,
      onSelect,
      EmojiGroups
    }
  }
})
</script>

<style lang="postcss" scoped>
.twemoji-picker {
  & >>> img {
    @apply inline;
  }

  & >>> #search-header > span {
    @apply box-content;
  }
}
</style>
