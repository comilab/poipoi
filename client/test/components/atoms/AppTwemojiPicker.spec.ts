import { render, RenderOptions } from '@testing-library/vue'
import { mocked } from 'ts-jest/utils'
import { defineComponent, ref } from 'nuxt-composition-api'
import { Ref } from '@vue/composition-api'
import EmojiAllData from '@kevinfaguiar/vue-twemoji-picker/emoji-data/ja/emoji-all-groups.json'
import EmojiGroups from '@kevinfaguiar/vue-twemoji-picker/emoji-data/emoji-groups.json'
import { useWindowSize } from '@vueuse/core'

import AppTwemojiPicker from '~/components/atoms/AppTwemojiPicker.vue'

jest.mock('@vueuse/core')

describe('components/atoms/AppTwemojiPicker', () => {
  let options: RenderOptions<AppTwemojiPicker>
  let props: any
  let twemojiPickerStub
  let twemojiPickerProps: any
  let twemojiPickerFireAddedFunction: Function
  let twemojiPickerFireAddedFunctionArg: string
  let windowWidth: Ref<number>

  beforeEach(() => {
    twemojiPickerStub = defineComponent({
      props: [
        'emojiData',
        'emojiGroups',
        'pickerWidth',
        'searchEmojisFeat',
        'emojiPickerDisabled',
        'emojiUnicodeAdded'
      ],
      setup (props, { emit }) {
        twemojiPickerProps = props
        twemojiPickerFireAddedFunction = () => emit('emojiUnicodeAdded', twemojiPickerFireAddedFunctionArg)
      },
      template: '<div><slot name="twemoji-picker-button"></slot></div>'
    })

    windowWidth = ref(1)
    mocked(useWindowSize).mockImplementation(() => {
      return {
        width: windowWidth,
        height: ref(0)
      }
    })

    props = {}

    options = {
      props,
      slots: {
        default: 'default-slot'
      },
      stubs: {
        TwemojiPicker: twemojiPickerStub
      }
    }
  })

  test('renders correctly', () => {
    const { container, getByText } = render(AppTwemojiPicker, options)

    expect(container).toMatchSnapshot()
    getByText('default-slot')
  })

  test('emojiGroups„ÅåÊ≠£Â∏∏„Å´Ê∏°„Åï„Çå„Å¶„ÅÑ„Çã', () => {
    render(AppTwemojiPicker, options)

    expect(twemojiPickerProps.emojiGroups).toBe(EmojiGroups)
  })

  test('searchEmojisFeat = true', () => {
    render(AppTwemojiPicker, options)

    expect(twemojiPickerProps.searchEmojisFeat).toBeTruthy()
  })

  describe('props', () => {
    describe('emojis', () => {
      test('ÂàùÊúüÂÄ§„Å™„ÇâÂÖ®ÁµµÊñáÂ≠ó„ÇíÊ∏°„Åô', () => {
        render(AppTwemojiPicker, options)

        expect(twemojiPickerProps.emojiData).toBe(EmojiAllData)
      })

      test('Á©∫ÈÖçÂàó„ÅåÊ∏°„Åï„Çå„ÅüÂ†¥Âêà„ÇÇÂÖ®ÁµµÊñáÂ≠ó„ÇíÊ∏°„Åô', () => {
        props.emojis = []
        render(AppTwemojiPicker, options)

        expect(twemojiPickerProps.emojiData).toBe(EmojiAllData)
      })

      test('Ê∏°„Åï„Çå„ÅüÁµµÊñáÂ≠ó„ÅßEmojiAllData„Çí„Éï„Ç£„É´„Çø„É™„É≥„Ç∞„Åô„Çã', () => {
        props.emojis = ['üç£']
        render(AppTwemojiPicker, options)

        const emojiData = twemojiPickerProps.emojiData
        expect(emojiData).toHaveLength(1)
        expect(emojiData[0].emojiList).toHaveLength(1)
        expect(emojiData[0].emojiList[0].unicode).toBe('üç£')
      })
    })

    describe('disabled', () => {
      test('ÂàùÊúüÂÄ§„ÅØfalse', () => {
        render(AppTwemojiPicker, options)

        expect(twemojiPickerProps.emojiPickerDisabled).toBeFalsy()
      })

      test('emojiPickerDisabled„Å´ÂÄ§„ÅåÊ∏°„Åï„Çå„Çã', () => {
        props.disabled = true
        render(AppTwemojiPicker, options)

        expect(twemojiPickerProps.emojiPickerDisabled).toBeTruthy()
      })
    })

    describe('pickerWidth', () => {
      test('„Ç¶„Ç£„É≥„Éâ„Ç¶ÂπÖ„Åå532px„ÇíË∂Ö„Åà„ÇãÂ†¥Âêà„ÅØ500„ÇíÊ∏°„Åô', () => {
        windowWidth.value = 533
        render(AppTwemojiPicker, options)

        expect(twemojiPickerProps.pickerWidth).toBe(500)
      })

      test('„Ç¶„Ç£„É≥„Éâ„Ç¶ÂπÖ„Åå532px‰ª•‰∏ã„ÅÆÂ†¥Âêà„ÅØ-32„Å´„Åó„ÅüÂÄ§„ÇíÊ∏°„Åô', () => {
        windowWidth.value = 531
        render(AppTwemojiPicker, options)

        expect(twemojiPickerProps.pickerWidth).toBe(499)
      })
    })

    describe('emit', () => {
      describe('select', () => {
        beforeEach(() => {
          twemojiPickerFireAddedFunctionArg = 'üç£'
        })

        test('emojiUnicodeAdded„Åï„Çå„ÅüÊôÇ„Å´Áô∫ÁÅ´', () => {
          const { emitted } = render(AppTwemojiPicker, options)
          twemojiPickerFireAddedFunction()

          expect(emitted().select).toHaveLength(1)
          expect(emitted().select[0]).toStrictEqual(['üç£'])
        })
      })
    })
  })
})
