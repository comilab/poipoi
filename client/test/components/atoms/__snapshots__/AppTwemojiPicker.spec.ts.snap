// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/atoms/AppTwemojiPicker renders correctly 1`] = `
<div>
  <div
    class="twemoji-picker"
  >
    default-slot
  </div>
</div>
`;
